cloud/profitbricks/profitbricks.py

```diff
@@ -0,0 +1,632 @@
+
+            if len(matched_instances) < len(instance_ids):
+                time.sleep(5)
+            else:
+                break
+
+        if wait_timeout <= time.time():
+            # waiting took too long
+            module.fail_json(msg = "wait for virtual machine state timeout on %s" % time.asctime())
+
+    return (changed)
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            datacenter=dict(),
+            name=dict(),
+            image=dict(),
+            cores=dict(default=2),
+            ram=dict(default=2048),
+            volume_size=dict(default=10),
+            bus=dict(default='VIRTIO'),
+            lan=dict(default=1),
+            count=dict(default=1),
+            auto_increment=dict(type='bool', default=True),
+            instance_ids=dict(),
+            subscription_user=dict(),
+            subscription_password=dict(),
+            location=dict(choices=LOCATIONS, default='us/las'),
+            assign_public_ip=dict(type='bool', default=False),
+            wait=dict(type='bool', default=True),
+            wait_timeout=dict(default=600),
+            remove_boot_volume=dict(type='bool', default=True),
+            state=dict(default='present'),
+        )
+    )
+
+    subscription_user = module.params.get('subscription_user')
+    subscription_password = module.params.get('subscription_password')
+    wait = module.params.get('wait')
+    wait_timeout = int(module.params.get('wait_timeout'))
+
+    profitbricks = ProfitBricksService(
+        username=subscription_user,
+        password=subscription_password)
+
+    state = module.params.get('state')
+
+    if state == 'absent':
+        if not module.params.get('datacenter'):
+            module.fail_json(msg='datacenter parameter is required for running or stopping machines.')
+
+        (changed) = remove_virtual_machine(module, profitbricks)
+
+        module.exit_json(
+            changed=changed)
+
+    elif state in ('running', 'stopped'):
+        if not module.params.get('datacenter'):
+            module.fail_json(msg='datacenter parameter is required for running or stopping machines.')
+        
+        (changed) = startstop_machine(module, profitbricks, state)
+
+        module.exit_json(
+            changed=changed)
+
+    elif state == 'present':
+        if not module.params.get('name'):
+            module.fail_json(msg='name parameter is required for new instance')
+        if not module.params.get('image'):
+            module.fail_json(msg='image parameter is required for new instance')
+        if not module.params.get('subscription_user'):
+            module.fail_json(msg='subscription_user parameter is required for new instance')
+        if not module.params.get('subscription_password'):
+            module.fail_json(msg='subscription_password parameter is required for new instance')
+
+        (machine_dict_array) = create_virtual_machine(module, profitbricks)
```

Comment:
not all paths capture exceptions and turn them into fail_json messages, for example create_virtual_machine here.